### PR TITLE
NEW: Add IDaRS models

### PR DESCRIPTION
**Pull request by simongraham**
- Add IDaRS models within the toolbox. For this, we add models for:
   - Tumour segmentation
   - MSI prediction
   - TP53 prediction
   - BRAF prediction
   - HM prediction
   - CIN prediction
   - CIMP prediction

- Add probability map function. This is useful for IDaRS visualisation.

### CODE REVIEW

File: tests/models/test_arch_idars.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#
+# The Original Code is Copyright (C) 2021, TIA Centre, University of Warwick
+# All rights reserved.
+# ***** END GPL LICENSE BLOCK *****
+
+"""Functional unit test package for IDARS."""
+
+import torch
+
+from tiatoolbox.models.architecture.idars import CNNIDaRS
```
Comment (shaneahmed) [inline tests/models/test_arch_idars.py:25]: why CNNIDaRS? why not just IDaRS?

File: tiatoolbox/models/architecture/idars.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+TRANSFORM = transforms.Compose(
+    [
+        transforms.ToTensor(),
+        transforms.Normalize(mean=[0.5, 0.5, 0.5], std=[0.1, 0.1, 0.1]),
+    ]
+)
+
+
+class CNNIDaRS(CNNModel):
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/idars.py:36]: I would suggest IDaRS.

File: setup.py [hunk 3]
```diff
@@ -32,6 +32,7 @@
     "zarr",
     "requests",
     "defusedxml",
+    "umap-learn",
```
Comment (shaneahmed) [inline setup.py:35]: do we need to add this to requirements files as well?

**Review by shaneahmed — COMMENTED** [review]
Thanks @simongraham  A couple of minor points added.

File: tiatoolbox/models/architecture/idars.py (hunk 2 above)
Comment (simongraham) [reply tiatoolbox/models/architecture/idars.py:36]: Without CNN portion?

File: setup.py (hunk 3 above)
Comment (simongraham) [reply setup.py:35]: I am not sure where this is being used - I merged from develop

File: tiatoolbox/models/architecture/idars.py (hunk 2 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/idars.py:36]: yes just name it `IDaRS`

File: setup.py (hunk 3 above)
Comment (shaneahmed) [reply setup.py:35]: ok I can see it is already in requirements files so ignore this comment.

File: tiatoolbox/models/architecture/idars.py (hunk 2 above)
Comment (simongraham) [reply tiatoolbox/models/architecture/idars.py]: Refer to this[ commit](https://github.com/TissueImageAnalytics/tiatoolbox/pull/214/commits/0589b8c774d5a76440e1d5f2ee3fae61a4ea8f49).

File: tests/models/test_arch_idars.py (hunk 1 above)
Comment (simongraham) [reply tests/models/test_arch_idars.py]: Refer to this [commit.](https://github.com/TissueImageAnalytics/tiatoolbox/pull/214/commits/0589b8c774d5a76440e1d5f2ee3fae61a4ea8f49)

File: tiatoolbox/models/engine/semantic_segmentor.py [hunk 4]
```diff
@@ -440,7 +440,7 @@ def __init__(
         self.pretrained_model = pretrained_model
         self.batch_size = batch_size
         self.num_loader_workers = num_loader_workers
-        self.num_postproc_workers = None
+        self.num_postproc_workers = num_postproc_workers
         self.verbose = verbose
         self.auto_generate_mask = auto_generate_mask
```
Comment (vqdang) [inline tiatoolbox/models/engine/semantic_segmentor.py:445]: @simongraham set to None

Comment (vqdang) [reply tiatoolbox/models/engine/semantic_segmentor.py:445]: We force it, always.  `num_postproc_workers` argument is for compatibility with subclasses (which may actually have post proc, and they will need to set it posthoc (overwriting) ). I dont think deepsource will complain.

**Review by vqdang — CHANGES_REQUESTED** [review]
Minor comment.

File: tiatoolbox/models/engine/semantic_segmentor.py (hunk 4 above)
Comment (simongraham) [reply tiatoolbox/models/engine/semantic_segmentor.py:445]: It does complain - it says unused argument if I do this

Comment (simongraham) [reply tiatoolbox/models/engine/semantic_segmentor.py:445]: You are welcome to try if you like and make a push

Comment (shaneahmed) [reply tiatoolbox/models/engine/semantic_segmentor.py:445]: @vqdang  https://deepsource.io/gh/TissueImageAnalytics/tiatoolbox/run/1b7bea27-6bb5-4cf5-b1e5-101652aec4aa/python/PYL-W0613

**Review by shaneahmed — APPROVED** [review]
Thanks @simongraham

**Pull request merged**